File: src/main/java/br/com/encrypt/Criptografia.java
package br.com.encrypt;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import javax.swing.JOptionPane;

import br.com.model.entities.classes.Usuario.Usuario;
import de.mkammerer.argon2.Argon2;
import de.mkammerer.argon2.Argon2Factory;
import de.mkammerer.argon2.Argon2Helper;

public class Criptografia {

    public static String argon(String senha) {
        try {
            String hash;
            Argon2 argon2 = Argon2Factory.create(Argon2Factory.Argon2Types.ARGON2id, 16, 32);
            var start = Instant.now();

            var iterations = Argon2Helper.findIterations(argon2, 1000, 65536, 1);

            System.out.println("\nOptimal number of iterations: " + iterations);

            hash = argon2.hash(iterations, 65536, 1, senha.toCharArray());

            System.out.println("\n" + hash);

            var end = Instant.now();
            System.out.println("Hashing took\n" + ChronoUnit.MILLIS.between(start, end) + "ms.");

            return hash;

        } catch (Exception e) {
            JOptionPane.showMessageDialog(null, e.getMessage(), "Error", JOptionPane.ERROR_MESSAGE);
            return e.getMessage();
        }
    }

    public static boolean verificar(Usuario usuario, String senha) {
        Argon2 argon2 = Argon2Factory.create(Argon2Factory.Argon2Types.ARGON2id, 16, 32);
        System.out.println("\nSenha do usuario: " + usuario.getSenha() + "\nSenha a ser verificada: " + senha);
        if (argon2.verify(usuario.getSenha(), senha.toCharArray())) {
            System.out.println("Tudo blz!");
            return true;
        } else {
            System.out.println("Deu ruim!\nA verificação dos Hash não batem.. =/");
            return false;
        }
    }
}

File: src/main/java/br/com/encrypt/Criptografia.kt
package br.com.encrypt

import br.com.model.usuario.Usuario
import de.mkammerer.argon2.Argon2Factory
import de.mkammerer.argon2.Argon2Helper
import kotlin.time.measureTime

object Criptografia {
    private var argon2 = Argon2Factory.create(Argon2Factory.Argon2Types.ARGON2id)
    private var parallelism = Runtime.getRuntime().availableProcessors()
    private const val MEMORYFREE = 65536
    private val iterations = Argon2Helper.findIterations(argon2, 500, MEMORYFREE, parallelism).also {
        println("Optimal number of iterations: $it")
    }

    @JvmStatic
    fun argon(senha: String): String {
        return run {
            println("\nSenha: \n$senha\n")
            val hash: String
            println("Demorou: ${
                measureTime {
                    hash = argon2.hash(iterations, MEMORYFREE, parallelism, senha.toCharArray())
                }
            } para gerar o Hash.\nHash:\n$hash")
            hash
        }
    }

    @JvmStatic
    fun verificar(usuario: Usuario, senha: String): Boolean {
        println("\nHash do usuario:${usuario.senha}\nSenha a ser verificada:$senha")
        return if (argon2.verify(usuario.senha, senha.toCharArray())) {
            println("Tudo blz!")
            true
        } else {
            println("Deu ruim!\nA verificação dos Hash não batem… =/")
            false
        }
    }
}
